File: benchmark.lua
-- Mock globals
CombatLogGetCurrentEventInfo = function() return nil, "UNIT_DIED", nil, nil, nil, nil, nil, "Creature-0-0-0-0-1234", "Random Mob", nil, nil, nil, nil, nil, nil end
strsub = string.sub
time = os.time
GetTime = os.clock

DUNGEONS = {
    [1417] = {
        name = "The Deadmines",
        cap = 20,
        bosses = {
            "Rhahk'Zor", "Sneed", "Gilnid", "Mr. Smite", "Cookie", "Captain Greenskin", "Edwin VanCleef"
        }
    }
}

currentRun = {
    zoneID = 1417,
    mobCount = 0,
    bossesKilled = {},
    done = false
}

addon = {}
function addon:AnnounceBossKill(bossName) end
function addon:CheckCompletion() end
function addon:UpdateProgressFrame() end
function addon:SaveRunState() end

local function run_benchmark()
    local start_time = os.clock()
    for i = 1, 1000000 do
        local _, subEvent, _, _, _, _, _, destGUID, destName, _, _, _, _, _, _ = CombatLogGetCurrentEventInfo()

        if subEvent == "UNIT_DIED" then
            if destGUID and (strsub(destGUID, 1, 8) == "Creature" or strsub(destGUID, 1, 7) == "Vehicle") then
                currentRun.mobCount = currentRun.mobCount + 1

                local dungeon = DUNGEONS[currentRun.zoneID]
                for _, bossName in ipairs(dungeon.bosses) do
                    if destName == bossName and not currentRun.bossesKilled[bossName] then
                        currentRun.bossesKilled[bossName] = time()
                        addon:AnnounceBossKill(bossName)
                    end
                end

                addon:CheckCompletion()
                addon:UpdateProgressFrame()
                addon:SaveRunState()
            end
        end
    end
    local end_time = os.clock()
    print("Baseline Time: " .. (end_time - start_time) .. "s")
end

run_benchmark()

File: benchmark.py
import time

class Run:
    def __init__(self):
        self.zoneID = 1417
        self.mobCount = 0
        self.bossesKilled = {}
        self.done = False

currentRun = Run()

DUNGEONS = {
    1417: {
        'name': "The Deadmines",
        'cap': 20,
        'bosses': [
            "Rhahk'Zor", "Sneed", "Gilnid", "Mr. Smite", "Cookie", "Captain Greenskin", "Edwin VanCleef"
        ]
    }
}

class Addon:
    def AnnounceBossKill(self, bossName): pass
    def CheckCompletion(self): pass
    def UpdateProgressFrame(self): pass
    def SaveRunState(self): pass

addon = Addon()

def run_benchmark():
    start_time = time.time()
    for _ in range(1000000):
        subEvent = "UNIT_DIED"
        destGUID = "Creature-0-0-0-0-1234"
        destName = "Random Mob"

        if subEvent == "UNIT_DIED":
            if destGUID and (destGUID[:8] == "Creature" or destGUID[:7] == "Vehicle"):
                currentRun.mobCount += 1

                dungeon = DUNGEONS[currentRun.zoneID]
                boss_killed = False
                for bossName in dungeon['bosses']:
                    if destName == bossName and not currentRun.bossesKilled.get(bossName):
                        currentRun.bossesKilled[bossName] = time.time()
                        addon.AnnounceBossKill(bossName)
                        boss_killed = True

                # BASELINE: always called
                addon.CheckCompletion()
                addon.UpdateProgressFrame()
                addon.SaveRunState()

    print(f"Baseline Time: {time.time() - start_time:.4f}s")

run_benchmark()

def run_optimized_benchmark():
    currentRun.mobCount = 0
    start_time = time.time()
    for _ in range(1000000):
        subEvent = "UNIT_DIED"
        destGUID = "Creature-0-0-0-0-1234"
        destName = "Random Mob"

        if subEvent == "UNIT_DIED":
            if destGUID and (destGUID[:8] == "Creature" or destGUID[:7] == "Vehicle"):
                currentRun.mobCount += 1

                dungeon = DUNGEONS[currentRun.zoneID]
                boss_killed = False
                for bossName in dungeon['bosses']:
                    if destName == bossName and not currentRun.bossesKilled.get(bossName):
                        currentRun.bossesKilled[bossName] = time.time()
                        addon.AnnounceBossKill(bossName)
                        boss_killed = True

                # OPTIMIZED: only called if boss killed
                if boss_killed:
                    addon.CheckCompletion()
                    addon.UpdateProgressFrame()
                    addon.SaveRunState()

    print(f"Optimized Time: {time.time() - start_time:.4f}s")

run_optimized_benchmark()
